File: csharp/src/AutomationFramework.Core/Models/Product.cs
using System.Text.Json.Serialization;

namespace AutomationFramework.Core.Models;

/// <summary>
/// A product in the catalogue.
/// </summary>
/// <remarks>
/// <c>Price</c> is a <see cref="decimal"/>, not a <c>double</c>. Money in binary floating point produces
/// assertions that fail by a cent for no visible reason, and there is no upside to it here.
/// <para>
/// <c>Specs</c> is populated only by <c>GET /products/{id}</c>; the list endpoint omits it.
/// </para>
/// </remarks>
public sealed record Product(
    [property: JsonPropertyName("id")] string Id,
    [property: JsonPropertyName("name")] string Name,
    [property: JsonPropertyName("description")] string? Description,
    [property: JsonPropertyName("price")] decimal Price,
    [property: JsonPropertyName("in_stock")] bool? InStock,
    [property: JsonPropertyName("is_location_offer")] bool? IsLocationOffer,
    [property: JsonPropertyName("is_rental")] bool? IsRental,
    [property: JsonPropertyName("is_eco_friendly")] bool? IsEcoFriendly,
    [property: JsonPropertyName("co2_rating")] string? Co2Rating,
    [property: JsonPropertyName("category")] Category? Category,
    [property: JsonPropertyName("brand")] Brand? Brand,
    [property: JsonPropertyName("product_image")] ProductImage? ProductImage,
    [property: JsonPropertyName("specs")] IReadOnlyList<ProductSpec>? Specs);

File: java/src/main/java/com/company/automation/models/Product.java
package com.company.automation.models;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonProperty;
import java.math.BigDecimal;

/**
 * A product in the catalogue.
 *
 * <p>{@code price} is a {@link BigDecimal}, not a {@code double}. Money in binary floating point
 * produces assertions that fail by a cent for no visible reason, and there is no upside to it here.
 */
@JsonIgnoreProperties(ignoreUnknown = true)
public record Product(
    @JsonProperty("id") String id,
    @JsonProperty("name") String name,
    @JsonProperty("description") String description,
    @JsonProperty("price") BigDecimal price,
    @JsonProperty("in_stock") Boolean inStock,
    @JsonProperty("is_location_offer") Boolean isLocationOffer,
    @JsonProperty("is_rental") Boolean isRental,
    @JsonProperty("is_eco_friendly") Boolean isEcoFriendly,
    @JsonProperty("co2_rating") String co2Rating,
    @JsonProperty("category") Category category,
    @JsonProperty("brand") Brand brand,
    @JsonProperty("product_image") ProductImage productImage) {}
